### PR TITLE
[MXNET-486] Create CPP test for concat MKLDNN operator

### PR DESCRIPTION
**Pull request by azai91**
## Description ##
Add test for MKLDNN concat forward / backward operator

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add test for MKLDNN concat forward / backward operator
- [x] Modify GetTestInputArrays / GetOutputInputArrays to support `num_inputs` / `dim` so that the inputs / outputs can be scaled (used to create larger ndarrays scaled by `num_inputs` on `dim`)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by azai91** [comment]
Does not appear like this causes regression as well:
```
ubuntu@ip-172-31-11-93:~/incubator-mxnet$ python example/image-classification/benchmark_score.py
/home/ubuntu/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Assertion failure at kmp_runtime.cpp(3767): __kmp_gtid_get_specific() == gtid.
INFO:root:network: resnet-50
INFO:root:device: cpu(0)
/home/ubuntu/incubator-mxnet/python/mxnet/module/base_module.py:66: UserWarning: Data provided by label_shapes don't match names specified by label_names ([] vs. ['softmax_label'])
  warnings.warn(msg)
[06:49:57] ../src/operator/nn/mkldnn/mkldnn_base.cc:72: Allocate 147456 bytes with malloc directly
[06:49:57] ../src/operator/nn/mkldnn/mkldnn_base.cc:72: Allocate 589824 bytes with malloc directly
[06:49:57] ../src/operator/nn/mkldnn/mkldnn_base.cc:72: Allocate 2359296 bytes with malloc directly
[06:49:57] ../src/operator/nn/mkldnn/mkldnn_base.cc:72: Allocate 9437184 bytes with malloc directly
INFO:root:batch size  1, image/sec: 17.726948
INFO:root:batch size  2, image/sec: 30.507246
INFO:root:batch size  4, image/sec: 40.782929
INFO:root:batch size  8, image/sec: 47.306150
INFO:root:batch size 16, image/sec: 49.509562
INFO:root:batch size 32, image/sec: 51.801888
```

**Comment by azai91** [comment]
@zheng-da please review when you have time.

File: CMakeLists.txt [hunk 1]
```diff
@@ -185,6 +185,8 @@ if(USE_VTUNE)
   list(APPEND mxnet_LINKER_LIBS dl)
 endif()
 
+set(USE_MKLDNN ON)
```
Comment (larroy) [inline CMakeLists.txt]: Why are we setting this flag on the makefile directly?

**Comment by azai91** [comment]
my bad. didnt mean to commit that. will revert.

File: src/operator/nn/concat.cc [hunk 2]
```diff
@@ -157,7 +157,19 @@ inline static bool BackwardConcatStorageType(const nnvm::NodeAttrs& attrs,
   return storage_type_assign(out_attrs, mxnet::kDefaultStorage,
                              dispatch_mode, wanted_mode);
 }
-
+#if MXNET_USE_MKLDNN == 1
+bool SupportMKLDNNConcat(const std::vector<NDArray> &arrs) {
+  for (auto arr : arrs) {
```
Comment (larroy) [inline src/operator/nn/concat.cc]: This is copying the objects adding overhead, why not use const auto& arr: arrs? or at least by reference?

Comment (larroy) [inline src/operator/nn/concat.cc]: This is making copies, why not use references?   auto& arr or even better a const reference if possible?

File: CMakeLists.txt (hunk 1 above)
Comment (larroy) [inline CMakeLists.txt]: Why are we setting this flag on the CMakefile directly?

File: tests/cpp/operator/mkldnn.cc [hunk 3]
```diff
@@ -160,6 +160,17 @@ static mkldnn::memory::primitive_desc GetMemPD(const TShape s, int dtype,
   return mkldnn::memory::primitive_desc(desc, CpuEngine::Get()->get_engine());
 }
 
+static mkldnn::memory::primitive_desc GetExpandedMemPD(
+    mkldnn::memory::primitive_desc pd, float num_input, int dim = 0) {
+  CHECK(dim < pd.desc().data.ndims);
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: Can we add an error message to this check with "<<" so we know why this invariant should hold?

File: CMakeLists.txt (hunk 1 above)
Comment (larroy) [inline CMakeLists.txt]: Why are we setting this flag on the CMakefile directly?

File: src/operator/nn/concat.cc (hunk 2 above)
Comment (larroy) [inline src/operator/nn/concat.cc]: Is making copying of the arrays intended on the loop? Why not reference?

File: tests/cpp/operator/mkldnn.cc [hunk 4]
```diff
@@ -446,20 +493,40 @@ OpAttrs GetSumBackwardsOp() {
...
   std::vector<nnvm::TShape> shapes = tas.shapes;
   std::vector<mkldnn::memory::primitive_desc> pds = tas.pds;
 
   std::vector<NDArrayAttrs> in_arrs;
   std::string desc;
+
+  int slice_amount = 1;
+  if (num_inputs != 0 && dim == 0)
+    slice_amount = num_inputs;
   for (auto shape : shapes) {
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: Is this copying intended?

File: tests/cpp/operator/mkldnn.cc [hunk 5]
```diff
@@ -604,6 +662,46 @@ std::vector<NDArrayAttrs> GetTestOutputArrays(const TShape &shape,
...
+  std::vector<nnvm::TShape> shapes; shapes = shapes_pds.shapes;
+  std::vector<mkldnn::memory::primitive_desc> pds = shapes_pds.pds;
+  for (auto shape : shapes) {
+    for (int dim = 0; dim < 5; dim++) {
+      for (int num_inputs = 2; num_inputs < 5; num_inputs++) {
+        if (shape.ndim() <= dim)
+          continue;
+        std::cout << "Extending " << shape << " dim " << dim << " and " << num_inputs << "num_inputs\n";
+        auto output_arrs = GetTestOutputArrays(shape, pds, num_inputs, dim);
+        for (auto out_arr : output_arrs) {
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: same

File: tests/cpp/operator/mkldnn.cc [hunk 6]
```diff
@@ -604,6 +662,46 @@ std::vector<NDArrayAttrs> GetTestOutputArrays(const TShape &shape,
...
+      }
+    }
+  }
+}
+
+TEST(MKLDNN_NDArray, GetTestOutputArraysConcat) {
+  auto shapes_pds = GetTestArrayShapes();
+  std::vector<nnvm::TShape> shapes; shapes = shapes_pds.shapes;
+  std::vector<mkldnn::memory::primitive_desc> pds = shapes_pds.pds;
+  for (auto shape : shapes) {
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: same

File: tests/cpp/operator/mkldnn.cc [hunk 7]
```diff
@@ -694,10 +852,10 @@ TEST(MKLDNN_NDArray, CopyFrom) {
 
   std::vector<NDArrayAttrs> in_arrs = GetTestInputArrays();
   for (auto in_arr : in_arrs) {
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: same

File: tests/cpp/operator/mkldnn.cc [hunk 8]
```diff
@@ -718,26 +877,61 @@ void TestOp(const OpAttrs &attrs, VerifyFunc verify_fn) {
...
         verify_fn(inputs, outputs);
       }
     }
   }
 
+  // cannot WriteInPlace concat operation
+  if (use_concat_outputs)
+    return;
+
   for (auto dispatch : dispatches) {
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: copies?

File: tests/python/unittest/test_loss.py [hunk 9]
```diff
@@ -209,7 +209,7 @@ def test_ctc_loss():
 
 @unittest.skip("flaky test. https://github.com/apache/incubator-mxnet/issues/8892")
 @with_seed(1234)
-def test_ctc_loss_train():
+def .test_ctc_loss_train():
```
Comment (larroy) [inline tests/python/unittest/test_loss.py]: is this change intended?

Comment (azai91) [reply tests/python/unittest/test_loss.py]: wow my stash was applied to this branch. will revert

File: tests/cpp/operator/mkldnn.cc [hunk 10]
```diff
@@ -446,20 +503,40 @@ OpAttrs GetSumBackwardsOp() {
...
   TestArrayShapes tas = GetTestArrayShapes();
   std::vector<nnvm::TShape> shapes = tas.shapes;
   std::vector<mkldnn::memory::primitive_desc> pds = tas.pds;
 
   std::vector<NDArrayAttrs> in_arrs;
   std::string desc;
+
+  int slice_amount = 1;
+  if (num_inputs != 0 && dim == 0)
+    slice_amount = num_inputs;
```
[the remark this answers is not in the document]
Comment (azai91) [reply tests/cpp/operator/mkldnn.cc]: the total size of the array needs to be divisible by the num_inputs. if we just subtract 2 (which we do currently) then the arrays will not always be divisible by the num of inputs (such as if num_inputs = 3)

File: tests/cpp/operator/mkldnn.cc [hunk 11]
```diff
@@ -708,7 +870,8 @@ TEST(MKLDNN_NDArray, CopyFrom) {
   }
 }
 
-void TestOp(const OpAttrs &attrs, VerifyFunc verify_fn) {
+void TestOp(const OpAttrs &attrs, VerifyFunc verify_fn,
+            bool use_concat_outputs = false, bool use_concat_inputs = false) {
```
[the remark this answers is not in the document]
Comment (azai91) [reply tests/cpp/operator/mkldnn.cc]: my issue with creating a separate TestOp for concat is that I would actually need to create two more. concat forward requires outputs to be larger and concat backwards requires inputs to be larger (and outputs to be relatively smaller).

[the remark this answers is not in the document]
Comment (zheng-da) [reply tests/cpp/operator/mkldnn.cc]: how about creating a single function for testing forward and backward of concat?

**Comment by azai91** [comment]
@pengzhao-intel

File: tests/cpp/operator/mkldnn.cc [hunk 12]
```diff
@@ -804,6 +1012,24 @@ TEST(IMPERATIVE, SumBackwardsOp) {
...
+  for (int num_inputs = 2; num_inputs < 4; num_inputs++) {
+    for (int dim = 0; dim < 5; dim++) {
+      OpAttrs attrs = GetConcatOp(num_inputs, dim);
+      TestConcatOp(attrs, VerifyConcatResult);
+    }
+  }
+}
+
+TEST(IMPERATIVE, ConcatBackwardsOp) {
+  for (int num_inputs = 2; num_inputs < 3; num_inputs++) {
```
Comment (zheng-da) [inline tests/cpp/operator/mkldnn.cc]: the same comment for here.
